File: backend/app/api/routes/analytics.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from typing import Optional

from app.core.database import get_db
from app.services import analytics_service

router = APIRouter(prefix="/analytics", tags=["analytics"])


@router.get("/salary-stats")
def salary_stats(
    country: str = Query(..., description="Country to analyse"),
    db: Session = Depends(get_db),
):
    """Min, max, avg salary for a given country."""
    return analytics_service.get_salary_stats_by_country(db, country)


@router.get("/salary-by-job-title")
def salary_by_job_title(
    country: str = Query(..., description="Country to analyse"),
    db: Session = Depends(get_db),
):
    """Average salary per job title within a country."""
    return analytics_service.get_avg_salary_by_job_title_in_country(db, country)


@router.get("/department-breakdown")
def department_breakdown(db: Session = Depends(get_db)):
    """Headcount, avg, min, max salary per department across the org."""
    return analytics_service.get_department_breakdown(db)


@router.get("/top-paid-roles")
def top_paid_roles(
    limit: int = Query(default=10, ge=1, le=30),
    db: Session = Depends(get_db),
):
    """Top N highest-paying job titles by average salary."""
    return analytics_service.get_top_paid_job_titles(db, limit)


@router.get("/salary-distribution")
def salary_distribution(
    country: Optional[str] = Query(default=None),
    db: Session = Depends(get_db),
):
    """Salary histogram bands — global or filtered by country."""
    return analytics_service.get_salary_distribution(db, country)


@router.get("/salary-percentiles")
def salary_percentiles(
    country: Optional[str] = Query(default=None),
    db: Session = Depends(get_db),
):
    """P25, P50, P75, P90 salary percentiles — global or by country."""
    return analytics_service.get_salary_percentile_bands(db, country)


@router.get("/headcount-by-country")
def headcount_by_country(db: Session = Depends(get_db)):
    """Total employees and avg salary per country."""
    return analytics_service.get_headcount_by_country(db)

File: frontend/src/api/analytics.ts
import { apiClient } from "./client";

export interface SalaryStats {
  country: string;
  min_salary: number;
  max_salary: number;
  avg_salary: number;
  headcount: number;
}

export interface JobTitleSalary {
  job_title: string;
  avg_salary: number;
  headcount: number;
}

export interface DepartmentBreakdown {
  department: string;
  headcount: number;
  avg_salary: number;
  min_salary: number;
  max_salary: number;
}

export interface SalaryBand {
  band: string;
  min: number;
  max: number;
  count: number;
}

export interface Percentiles {
  p25: number;
  p50: number;
  p75: number;
  p90: number;
  total_employees: number;
}

export interface CountryHeadcount {
  country: string;
  headcount: number;
  avg_salary: number;
}

export const getSalaryStats = async (country: string) => {
  const { data } = await apiClient.get<SalaryStats>("/analytics/salary-stats", {
    params: { country },
  });
  return data;
};

export const getSalaryByJobTitle = async (country: string) => {
  const { data } = await apiClient.get<JobTitleSalary[]>(
    "/analytics/salary-by-job-title",
    { params: { country } }
  );
  return data;
};

export const getDepartmentBreakdown = async () => {
  const { data } = await apiClient.get<DepartmentBreakdown[]>(
    "/analytics/department-breakdown"
  );
  return data;
};

export const getTopPaidRoles = async (limit = 10) => {
  const { data } = await apiClient.get<JobTitleSalary[]>(
    "/analytics/top-paid-roles",
    { params: { limit } }
  );
  return data;
};

export const getSalaryDistribution = async (country?: string) => {
  const { data } = await apiClient.get<SalaryBand[]>(
    "/analytics/salary-distribution",
    { params: country ? { country } : {} }
  );
  return data;
};

export const getSalaryPercentiles = async (country?: string) => {
  const { data } = await apiClient.get<Percentiles>(
    "/analytics/salary-percentiles",
    { params: country ? { country } : {} }
  );
  return data;
};

export const getHeadcountByCountry = async () => {
  const { data } = await apiClient.get<CountryHeadcount[]>(
    "/analytics/headcount-by-country"
  );
  return data;
};